[ClangImporter] Fix indirection mismatch surrounding swift_newtype

milseman opened the pull request:
Previously, we did not properly handle levels of indirection for
swift_newtype-ed typedefs of pointer types. We imported them in a way
that tried to present the value semantics of the pointee rather than
of the pointer. We then tried (sometimes incorrectly) to detect and
fix this up during SILGen.

Instead, model with the value semantics of the pointer itself. SILGen
can then be simplified to just pass swift_newtypes the same as any
other struct: directly for non-mutating and indirectly for mutating
(i.e. inout self). Tests added.

<!-- What's in this pull request? -->
<!-- If this pull request resolves any bugs in the Swift bug tracker, provide a link: -->
<rdar://problem/26899737>

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/master/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

huonw commented on lib/ClangImporter/ImportDecl.cpp:
```diff
@@ -5256,6 +5256,29 @@ Decl *SwiftDeclConverter::importGlobalAsInitializer(
+  auto clangDecl = structDecl->getClangDecl();
+  if (!clangDecl || !isa<clang::TypedefNameDecl>(clangDecl))
+    return false;
+  auto clangTypedefDecl = cast<clang::TypedefNameDecl>(clangDecl);
```
The previous 4 lines here could possibly be something like:

```cpp
auto clangTypedefDecl = dyn_cast_or_null<clang::TypedefNameDecl>(structDecl->getClangDecl());
if (!clangTypeDefDecl)
    return false;
```

milseman replied:
Thanks for the catch! One day I'll be able to use C++17.

milseman commented:
@swift-ci please test

milseman commented:
Failure in decl/protocol/conforms/fixit_stub_editor.swift seems unrelated.

milseman commented:
@swift-ci please test OS X Platform

DougGregor commented:
Ahhh, thank you!

milseman commented:
@swift-ci please smoke test OS X Platform

milseman commented:
Smoke test should fail now that it fails ast-verifier. Double checking.

milseman commented:
@swift-ci please smoke test OS X Platform

milseman committed:
[Clang Importer] Account for indirection when importing swift_newtypes

Previously, we did not properly handle levels of indirection for
swift_newtype-ed typedefs of pointer types. We imported them in a way
that tried to present the value semantics of the pointee rather than
of the pointer. We then tried (sometimes incorrectly) to detect and
fix this up during SILGen.

Instead, model with the value semantics of the pointer itself. SILGen
can then be simplified to just pass swift_newtypes the same as any
other struct: directly for non-mutating and indirectly for mutating
(i.e. inout self). Tests added.

milseman commented:
@DougGregor @jrose-apple completely new implementation.

milseman commented:
@swift-ci please test

milseman commented on lib/ClangImporter/ImportDecl.cpp:
```diff
@@ -5285,16 +5285,30 @@ Decl *SwiftDeclConverter::importGlobalAsMethod(
+        if (auto clangDCTy = dyn_cast_or_null<clang::TypedefNameDecl>(
+                nominalTypeDecl->getClangDecl()))
+          if (auto newtypeAttr = getSwiftNewtypeAttr(clangDCTy, getVersion()))
+            if (clangDCTy->getUnderlyingType().getCanonicalType() !=
```
The choice is either use `getUnderlyingType()` or `getTypeForDecl()`. The former returns a `QualType` and the latter a `const Type *`.

I'm not that familiar with the assumptions ingrained in Clang's in-memory representation of types. If the latter is always canonical and unqualified and using that is more clear, let me know. Otherwise using QualType's `getCanonicalType()` appears to me to be the more likely safe choice.

DougGregor replied:
Oh, ignore me it's a `Decl`, not a `QualType`.

milseman committed:
[tests] Don't check for alignment in swift_newtype test (32bit)

The alignment of pointers in the swift_newtype test is irrelevant, so
don't strictly check them.

milseman commented:
@swift-ci please test

milseman commented:
@swift-ci please test
